File: api/downloads.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"

	real_debrid "github.com/sushydev/real_debrid_go"
)

type Download struct {
	ID        string `json:"id"`
	Filename  string `json:"filename"`
	MimeType  string `json:"mimeType"`
	Filesize  int64  `json:"filesize"`
	Link      string `json:"link"`
	Host      string `json:"host"`
	Chunks    int    `json:"chunks"`
	Crc       int    `json:"crc"`
	Download  string `json:"download"`
	Generated string `json:"generated"`
	Type      string `json:"type,omitempty"`
}

// GetDownloads returns downloads list with total count from X-Total-Count header
func GetDownloads(client *real_debrid.Client, limit uint, page uint, offset *uint) ([]*Download, int, error) {
	url := client.GetUrl("/downloads")
	q := url.Query()
	if limit > 0 {
		q.Add("limit", strconv.Itoa(int(limit)))
	}
	if page > 0 {
		q.Add("page", strconv.Itoa(int(page)))
	}
	if offset != nil {
		q.Add("offset", strconv.Itoa(int(*offset)))
	}
	url.RawQuery = q.Encode()
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, 0, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, 0, err
	}
	var items []*Download
	if err := json.NewDecoder(resp.Body).Decode(&items); err != nil {
		return nil, 0, err
	}
	totalCountHeader := resp.Header.Get("X-Total-Count")
	if totalCountHeader == "" {
		return items, 0, nil
	}
	totalCount, err := strconv.Atoi(totalCountHeader)
	if err != nil {
		return nil, 0, fmt.Errorf("invalid X-Total-Count: %w", err)
	}
	return items, totalCount, nil
}

func DeleteDownload(client *real_debrid.Client, id string) error {
	url := client.GetUrl("/downloads/delete/" + id)
	req, err := http.NewRequest("DELETE", url.String(), nil)
	if err != nil {
		return err
	}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return client.HandleResponseCode(resp, 204)
}

File: lib/real_debrid/api/downloads.ex
defmodule RealDebrid.Api.Downloads do
  @moduledoc """
  Downloads management API functions.
  """

  alias RealDebrid.Client

  @type download :: %{
          id: String.t(),
          filename: String.t(),
          mime_type: String.t(),
          filesize: integer(),
          link: String.t(),
          host: String.t(),
          chunks: integer(),
          crc: integer(),
          download: String.t(),
          generated: String.t(),
          type: String.t() | nil
        }

  @doc """
  Get downloads list with total count.

  ## Parameters

    - `client` - The Real-Debrid client
    - `opts` - Optional keyword list:
      - `:limit` - Number of results per page
      - `:page` - Page number
      - `:offset` - Offset for results

  ## Returns

    - `{:ok, %{downloads: downloads, total_count: count}}` on success
    - `{:error, reason}` on failure
  """
  @spec get(Client.t(), keyword()) :: {:ok, %{downloads: [download()], total_count: integer()}} | {:error, term()}
  def get(%Client{} = client, opts \\ []) do
    params =
      []
      |> maybe_add_param(:limit, Keyword.get(opts, :limit))
      |> maybe_add_param(:page, Keyword.get(opts, :page))
      |> maybe_add_param(:offset, Keyword.get(opts, :offset))
      |> Map.new()

    case Client.get(client, "/downloads", params: params) do
      {:ok, body, headers} ->
        downloads = Enum.map(body, &parse_download/1)

        total_count =
          headers
          |> get_header("x-total-count")
          |> parse_integer(0)

        {:ok, %{downloads: downloads, total_count: total_count}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Delete a download by its ID.

  ## Returns

    - `:ok` on success
    - `{:error, reason}` on failure
  """
  @spec delete(Client.t(), String.t()) :: :ok | {:error, term()}
  def delete(%Client{} = client, id) do
    Client.delete(client, "/downloads/delete/#{id}")
  end

  defp parse_download(data) do
    %{
      id: data["id"],
      filename: data["filename"],
      mime_type: data["mimeType"],
      filesize: data["filesize"],
      link: data["link"],
      host: data["host"],
      chunks: data["chunks"],
      crc: data["crc"],
      download: data["download"],
      generated: data["generated"],
      type: data["type"]
    }
  end

  defp maybe_add_param(params, _key, nil), do: params
  defp maybe_add_param(params, key, value), do: [{key, value} | params]

  defp get_header(headers, key) do
    case List.keyfind(headers, key, 0) do
      {_, value} -> value
      nil -> nil
    end
  end

  defp parse_integer(nil, default), do: default
  defp parse_integer(value, default) when is_binary(value) do
    case Integer.parse(value) do
      {int, _} -> int
      :error -> default
    end
  end
  defp parse_integer(value, _default) when is_integer(value), do: value
end
